### PR TITLE
Source token signing keys from the database

### PR DESCRIPTION
**Pull request by sethvargo**
Fixes https://github.com/google/exposure-notifications-verification-server/issues/1569

**Release Note**

```release-note
**Potentially breaking!** Source token signing keys from the database. This completes the move of system token signing keys from environment variables to the database. This change attempts to be backward compatible, but server are encouraged to test changes in an isolated environment before upgrading production systems.
```

/assign @mikehelmick

### CODE REVIEW

File: pkg/controller/rotation/handle_rotate.go [hunk 1]
```diff
@@ -65,19 +66,19 @@ func (c *Controller) HandleRotate() http.Handler {
...
-			if err != nil {
+			if err != nil && !database.IsNotFound(err) {
 				merr = multierror.Append(merr, fmt.Errorf("failed to lookup existing signing key: %w", err))
 				result = observability.ResultError("FAILED")
 				return
 			}
-			if age, max := time.Now().UTC().Sub(existing.CreatedAt), c.config.TokenSigningKeyMaxAge; age < max {
-				logger.Debugw("token signing key does not require rotation", "age", age, "max", max)
-				return
+			if existing != nil {
```
Comment (sethvargo) [inline pkg/controller/rotation/handle_rotate.go:74]: This allows the rotator to create the first key if it does not exist

File: internal/routes/apiserver.go [hunk 2]
```diff
@@ -49,6 +49,23 @@ func APIServer(
...
+	if _, err := db.ActiveTokenSigningKey(); err != nil {
+		if !database.IsNotFound(err) {
+			return nil, closer, fmt.Errorf("failed to lookup active token signing key: %w", err)
+		}
+
+		tokenSignerTyp, ok := tokenSigner.(keys.SigningKeyManager)
+		if !ok {
+			return nil, closer, fmt.Errorf("token signing key manager is not a signing key manager (is %T)", tokenSigner)
+		}
+		if _, err := db.RotateTokenSigningKey(ctx, tokenSignerTyp, cfg.TokenSigning.ParentKeyName(), database.System); err != nil {
```
Comment (mikehelmick) [inline internal/routes/apiserver.go:64]: I think RotateTokenSigningKey needs some kind of locking mechanism (distributed).

It's possible that multiple container instances would try and do this at the same time.

Comment (sethvargo) [reply internal/routes/apiserver.go]: Chatted about this sync. We're going to allow the rotation service to bootstrap the initial key.

**Comment by sethvargo** [comment]
@mikehelmick updated, PTAL

File: cmd/server/assets/admin/info.html [hunk 3]
```diff
@@ -19,8 +19,23 @@
           <dt>Build ID</dt>
           <dd>{{.buildID}}</dd>
 
-          <dt>Build Tag</dt>
+          <dt>Build tag</dt>
           <dd>{{.buildTag}}</dd>
+
+          <dt>Token signing key</dt>
```
Comment (mikehelmick) [inline cmd/server/assets/admin/info.html:25]: can we display the allowed keys as well - there should be a 24h + overlap where more than one key is valid?

File: pkg/config/token_signing_config.go [hunk 4]
```diff
@@ -25,33 +26,79 @@ import (
...
 
-	TokenSigningKeys   []string `env:"TOKEN_SIGNING_KEY, required"`
+	// TokenSigningKeys is the parent token signing key (not the actual signing
+	// version). It is an array for backwards-compatibility, but in practice it
+	// should only have one element.
+	//
+	// Previously it was a list of all possible signing key versions, but those
+	// have moved into the database.
+	//
+	// TODO(sethvargo): Convert to string in 0.22+.
```
Comment (mikehelmick) [inline pkg/config/token_signing_config.go:38]: we'll have to tackle the terraform changes as well...

File: cmd/server/assets/admin/info.html (hunk 3 above)
Comment (sethvargo) [reply cmd/server/assets/admin/info.html]: Done

File: pkg/config/token_signing_config.go (hunk 4 above)
Comment (sethvargo) [reply pkg/config/token_signing_config.go:38]: They aren't actually that bad. This is the only change.

**Comment by sethvargo** [comment]
@mikehelmick PTAL

**Review by mikehelmick — APPROVED** [review]

**Pull request merged**